mayastor/src/bdev/dev/malloc.rs

```diff
@@ -0,0 +1,171 @@
+//! heap. IOW, you must ensure you do not run out of huge pages while using
+//! this.
+use crate::{
+    bdev::util::uri,
+    nexus_uri::{
+        NexusBdevError,
+        {self},
+    },
+};
+use std::{collections::HashMap, convert::TryFrom};
+use url::Url;
+
+use async_trait::async_trait;
+pub struct Malloc {
+    /// the name of the bdev we created, this is equal to the URI path minus
+    /// the leading '/'
+    name: String,
+    /// the number of blocks the device should have
+    num_blocks: u64,
+    /// the size of a single block if no blk_size is given we default to 512
+    blk_size: u32,
+}
+use crate::{
+    bdev::{CreateDestroy, GetName},
+    core::Bdev,
+    ffihelper::{cb_arg, done_errno_cb, ErrnoResult, IntoCString},
+};
+use futures::channel::oneshot;
+use nix::errno::Errno;
+use snafu::ResultExt;
+use spdk_sys::delete_malloc_disk;
+
+impl TryFrom<&Url> for Malloc {
+    type Error = NexusBdevError;
+
+    fn try_from(uri: &Url) -> Result<Self, Self::Error> {
+        let segments = uri::segments(uri);
+        if segments.is_empty() {
+            return Err(NexusBdevError::UriInvalid {
+                uri: uri.to_string(),
+                message: "no path segments".to_string(),
+            });
+        }
+
+        let mut parameters: HashMap<String, String> =
+            uri.query_pairs().into_owned().collect();
+
+        let blk_size: u32 = if let Some(value) = parameters.remove("blk_size") {
+            value.parse().context(nexus_uri::IntParamParseError {
+                uri: uri.to_string(),
+                parameter: String::from("blk_size"),
+            })?
+        } else {
+            512
+        };
+
+        if blk_size != 512 && blk_size != 4096 {
+            return Err(NexusBdevError::UriInvalid {
+                uri: uri.to_string(),
+                message:
+                    "invalid blk_size specified must be one of 512 or 4096"
+                        .to_string(),
+            });
+        }
+
+        let size: u32 = if let Some(value) = parameters.remove("size_mb") {
+            value.parse().context(nexus_uri::IntParamParseError {
+                uri: uri.to_string(),
+                parameter: String::from("size_mb"),
+            })?
+        } else {
+            0
+        };
+
+        let num_blocks: u32 =
+            if let Some(value) = parameters.remove("num_blocks") {
+                value.parse().context(nexus_uri::IntParamParseError {
+                    uri: uri.to_string(),
+                    parameter: String::from("blk_size"),
+                })?
+            } else {
+                0
+            };
+
+        if size != 0 && num_blocks != 0 {
+            return Err(NexusBdevError::UriInvalid {
+                uri: uri.to_string(),
+                message: "conflicting parameters num_blocks and size_mb are mutually exclusive"
+                    .to_string(),
+            });
+        }
+
```

Comment:
Just a minor note.  the Implementations for the other types check and issue a warning if there are any unused parameters in the URI